### ISSUE TITLE
minimize number of types

### ISSUE THREAD

**Issue by nsheff** [pepkit/looper#102]
instead of a different type for each arg, we should have only a few types (that take parameters if necessary), and then use these paramterized types repeatedly.

**Comment by michalstolarczyk** [comment]
Since the `type` argument has to be a function we need to use the `partial` function to parametrize the defined types, like so:

```python
run_subparser.add_argument(
            "--lump", type=partial(range, min=1, max=2), default=None,
            help="Maximum total input file size for a lump/batch of commands "
                 "in a single job (in GB)")
```

**Comment by nsheff** [comment]
or, could you have the `range` function return a function?

**Comment by michalstolarczyk** [comment]
Sure. Do you think it is better, though?

It would just save the `partial` word in the `add_argument` statement

**Comment by nsheff** [comment]
I guess it just eliminates the cognitive burden of figuring out the partial function for something kind of simple... maybe @vreuter has a preference.

**Comment by vreuter** [comment]
Redefining a builtin function feels more atypical than using a partial function, especially since this is in the context of a pretty long function, so the meaning of the new definition would apply across a larger scope.

Separately, though, for what it looks like is the intended behavior, I think the partial should be `partial(range, 1, 2)`; I think builtin `range` takes only positional args.

I think maybe I don't know the intent, though; it looks like the value for `type`, once invoked, would always be `[1]`? In that case, a way to avoid both `partial` and a `range` redefinition would be something like `type=(lambda: [1])`

**Comment by nsheff** [comment]
ok, you're actually brining up a separate point:

@michalstolarczyk  don't use the name `range` to define your function since that's already a built-in python function.  use `html_range` or something else...

@vreuter the intent wasn't to override an existing function; it's supposed to be our own function...

**Comment by vreuter** [comment]
Wouldn't that look something like this?
```
def html_range():
    return [1]
...
...add_argument(..., type=html_range)
```

That seems more confusing than just `...add_argument(type=(lambda: [1]))`.

**Comment by nsheff** [comment]
No `html_range` is more complicated than that:

https://github.com/pepkit/looper/blob/44ba0a12345fc8c7db851b3bb4152ed6ab3819a7/looper/parser_types.py#L8-L20

**Comment by vreuter** [comment]
Ah OK yes then _definitely_ I think that should be
- [ ] renamed
- [ ] used wherever relevant as the argument to the `type` parameter in `add_argument`

**Comment by nsheff** [comment]
Agreed. 

But the original question asked which would be better: using `partial` or having `html_range` return a function?

**Comment by vreuter** [comment]
> Agreed.
> 
> But the original question asked which would be better: using `partial` or having `html_range` return a function?

TLDR -- I don't have a strong preference then.

Maybe it depends on where and how often `html_range` is and will be used. Specifically, the function return seems like it may be preferably if the specific kind of case here is the most frequent. If our own code will often be invoking it, though, maybe the actual value as the result is more reasonable. It feels nice--when possible--to avoid a bunch of repeated parenthesis pairs like `html_range(..., min=1, max=2)()` since the relative infrequency of partial application is likely to cause more pause when a code reader comes across it.

**Comment by nsheff** [comment]
you wouldn't need those empty parents, you'd just say: `html_range(min=1, max=2)` and have it return a function.

**Comment by vreuter** [comment]
Right, that's why the hypothesized frequency distribution over use cases matters and I qualified with this:
> If our own code will often be invoking it, though, maybe the actual value as the result is more reasonable

**Comment by nsheff** [comment]
I understand now.

**Comment by nsheff** [comment]
@michalstolarczyk  let's go with the function return approach please.
